pkg/threescale/backend/backend.go

```diff
@@ -269,6 +484,48 @@ func (a *Application) updateUnlimitedCounter(metric string, incrementBy int) {
 	a.UnlimitedCounter[metric] = current + incrementBy
 }
 
+// adjustLocalState assumes that we have a new remote state (set on a) fetched from 3scale and modifies local state based
+// on the state obtained during cache flushing
+func (a *Application) adjustLocalState(flushingState cachedApp, remoteState LimitCounter) *Application {
+	a.RemoteState = remoteState
+
+	for metric := range a.LimitCounter {
+		lowestPeriod := a.LimitCounter.getLowestKnownPeriodForMetric(metric)
+
+		if lowestPeriod == nil {
+			continue
+		}
+
+		snapped, ok := flushingState.snapshot.LimitCounter[metric][*lowestPeriod]
+		if !ok {
+			continue
+		}
+
+		latest, ok := a.RemoteState[metric][*lowestPeriod]
+		if !ok {
+			continue
+		}
+
+		current := a.LimitCounter[metric][*lowestPeriod].current
+
+		var reportedHits int
+		delta := flushingState.deltas[metric]
+		if !flushingState.reportingErr {
+			reportedHits = delta
+		}
+		updated := current + latest.current - (snapped.current - delta) - reportedHits
+		a.LimitCounter[metric][*lowestPeriod].current = updated
+		a.LimitCounter[metric][*lowestPeriod].max = a.RemoteState[metric][*lowestPeriod].max
+	}
+
+	if !flushingState.reportingErr {
+		// reset the counter if we have reported already
+		a.UnlimitedCounter = make(UnlimitedCounter)
```

Comment:
The same issue here as elsewhere when we clear this. Unless I missed something, there is a window of time in which the app is unlocked and this could potentially receive new hits, but we would be losing them here.